gadget/raw.go

```diff
@@ -0,0 +1,295 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package gadget
+
+import (
+	"bytes"
+	"crypto"
+	_ "crypto/sha1"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/osutil"
+)
+
+// RawStructureWriter implements support for writing raw (bare) structures.
+type RawStructureWriter struct {
+	rootDir string
+	ps      *PositionedStructure
+}
+
+// NewRawStructureWriter returns a writer for given structure, that will load
+// the structure content data from provided root directory.
+func NewRawStructureWriter(rootDir string, ps *PositionedStructure) (*RawStructureWriter, error) {
+	if !ps.IsBare() {
+		return nil, fmt.Errorf("structure %v is not bare", ps)
+	}
+	rw := &RawStructureWriter{
+		rootDir: rootDir,
+		ps:      ps,
+	}
+	return rw, nil
+}
+
+// writeRawStream writes the input stream in that corresponds to provided
+// positioned content. The number of bytes read from input stream must match
+// exactly the declared size of positioned content entry.
+func writeRawStream(out io.WriteSeeker, pc *PositionedContent, in io.Reader) error {
+	if _, err := out.Seek(int64(pc.StartOffset), io.SeekStart); err != nil {
```

Comment:
why isn't `pc.StartOffset` already an int64?

Reply:
It's gadget.Size which is `type Size uint64`